File: cpp/lib/886.PossibleBipartition.cpp
// 886. Possible Bipartition

#include "886.PossibleBipartition.h"


bool LeetCode886PossibleBipartition::possibleBipartition(int N, vector<vector<int>>& dislikes)
{
    if (N < 3 || dislikes.size() < 2) return true;

    auto nodes = GenNodes_(N, dislikes);

    for (int i = 1; i < (*nodes).size(); ++i) {
        if ((*nodes)[i].color == Color::NONE && !IsBipartition_(i, Color::WHITE, *nodes)) return false;
    }

    delete nodes;
    return true;
}

bool LeetCode886PossibleBipartition::IsBipartition_(int current, Color assignColor, vector<Node>& nodes) {
    if (nodes[current].color == Color::NONE) nodes[current].color = assignColor;

    auto nextColor = (nodes[current].color == Color::RED)? Color::WHITE : Color::RED;

    for (auto& next: nodes[current].Nexts) {
        if (!IsValidNeighbor_(next, nodes[current].color, nextColor, nodes)) return false;
    }

    return true;
}

bool LeetCode886PossibleBipartition::IsValidNeighbor_(int neighber, Color currentColor, Color assignColor, vector<Node>& nodes) {
    return (nodes[neighber].color == Color::NONE && IsBipartition_(neighber, assignColor, nodes)) || nodes[neighber].color != currentColor;
}

vector<Node>* LeetCode886PossibleBipartition::GenNodes_(int N, vector<vector<int>>& dislikes) {
    vector<Node> *nodes = new vector<Node>(N+1);

    for (auto& edge: dislikes) {
        auto i = edge[0];
        auto j = edge[1];
        (*nodes)[i].Nexts.push_back(j);
        (*nodes)[j].Nexts.push_back(i);
    }

    return nodes;
}

File: csharp/src/886_PossibleBipartition.cs
using System.Collections.Generic;

/*
LeetCode: 886. Possible Bipartition
*/

public class LeetCodePossibleBipartition {
	private enum COLOR
	{
		NONE,
		WHITE,
		BLACK,
	}
	private class Node
	{
		public COLOR color = COLOR.NONE;
		public List<int> nexts = new List<int>();
	}

	public bool PossibleBipartition(
		int N, int[][] dislikes)
    {
		if (N == 0 || dislikes == null || dislikes.Length == 0)
			return true;
		var nodes = _GenGraph(N, dislikes);
		for (int id = 1; id <= N; ++id)
			if (nodes[id].color == COLOR.NONE &&
				!_IsBipartition(id, nodes, COLOR.WHITE))
				return false;

		return true;
	}

	private Node[] _GenGraph(int numNodes, int[][] edges)
	{
		var nodes = new Node[numNodes+1];
		for (int i = 0; i < nodes.Length; ++i)
			nodes[i] = new Node();
		foreach (var edge in edges)
		{
			var i = edge[0];
			var j = edge[1];
			nodes[i].nexts.Add(j);
			nodes[j].nexts.Add(i);
		}
		return nodes;
	}

	private bool _IsBipartition(int id, Node[] nodes, COLOR assignedColor)
	{
		nodes[id].color = assignedColor;
		var nextColor = assignedColor == COLOR.WHITE? COLOR.BLACK : COLOR.WHITE;
		foreach (var next in nodes[id].nexts)
		{
			if (nodes[next].color == COLOR.NONE &&
				!_IsBipartition(next, nodes, nextColor))
				return false;
			else if (nodes[next].color != nextColor)
				return false;
		}
		return true;
	}
}